File: python/core/game_loop.py
from enum import Enum, auto
from typing import Callable
import time


class GameState(Enum):
    PAUSED = auto()
    RUNNING = auto()
    GAME_OVER = auto()


class GameLoop:
    TICK = 0.15  # seconds per move

    def __init__(self, update: Callable[[float], None]) -> None:
        self.update = update
        self.accumulator = 0.0
        self.last_time = 0.0
        self.state = GameState.PAUSED
        self.stopped = False

    def start(self) -> None:
        self.last_time = time.perf_counter()
        self.state = GameState.RUNNING
        self.stopped = False
        while not self.stopped and self.state != GameState.GAME_OVER:
            self.tick()

    def stop(self) -> None:
        self.stopped = True

    def toggle_pause(self) -> None:
        if self.state == GameState.RUNNING:
            self.state = GameState.PAUSED
        else:
            self.state = GameState.RUNNING
            self.last_time = time.perf_counter()

    def tick(self) -> None:
        if self.stopped:
            return
        now = time.perf_counter()
        delta = now - self.last_time
        self.last_time = now
        if self.state == GameState.RUNNING:
            self.accumulator += delta
            while self.accumulator >= self.TICK and self.state == GameState.RUNNING:
                self.update(self.TICK)
                self.accumulator -= self.TICK
        time.sleep(0.01)

File: src/core/GameLoop.ts
export enum GameState {
  PAUSED,
  RUNNING,
  GAME_OVER,
}

export type GameLoopEventType =
  | 'tick'
  | 'pause'
  | 'resume'
  | 'gameover';

export class GameLoop extends EventTarget {
  static TICK = 150; // ms per move
  private accumulator = 0;
  private lastTime = 0;
  state: GameState = GameState.PAUSED;
  private frame = 0;
  private stopped = false;

  constructor(private update: (dt: number) => void) {
    super();
  }

  on(type: GameLoopEventType, listener: EventListenerOrEventListenerObject) {
    this.addEventListener(type, listener);
  }

  emit(type: GameLoopEventType) {
    this.dispatchEvent(new Event(type));
  }

  start() {
    this.lastTime = performance.now();
    this.state = GameState.RUNNING;
    this.emit('resume');
    this.stopped = false;
    this.frame = requestAnimationFrame(this.tick);
  }

  stop() {
    this.stopped = true;
    if (typeof cancelAnimationFrame !== 'undefined') {
      cancelAnimationFrame(this.frame);
    }
  }

  togglePause() {
    if (this.state === GameState.RUNNING) {
      this.state = GameState.PAUSED;
      this.emit('pause');
    } else {
      this.state = GameState.RUNNING;
      this.lastTime = performance.now();
      this.emit('resume');
    }
  }

  gameOver() {
    this.state = GameState.GAME_OVER;
    this.emit('gameover');
    this.stop();
  }

  private tick = (time: number) => {
    if (this.stopped) {
      return;
    }
    const delta = time - this.lastTime;
    this.lastTime = time;
    if (this.state === GameState.RUNNING) {
      this.accumulator += delta;
      while (
        this.accumulator >= GameLoop.TICK &&
        this.state === GameState.RUNNING
      ) {
        this.update(GameLoop.TICK / 1000);
        this.emit('tick');
        this.accumulator -= GameLoop.TICK;
      }
    }
    if (!this.stopped) {
      this.frame = requestAnimationFrame(this.tick);
    }
  };
}
